src/mlpack/methods/lmnn/lmnn_function_impl.hpp

```diff
@@ -439,28 +531,46 @@ double LMNNFunction<MetricType>::EvaluateWithGradient(
         // Calculate cost due to {data point, target neighbors, impostors}
         // triplets.
         double eval = 0;
-        if ((iteration - 1 % range == 0) && (iteration - 1 != 0))
+        if (transformationOld.n_elem != 0)
         {
-          eval = metric.Evaluate(transformedDataset.col(i),
-                    transformedDataset.col(targetNeighbors(j, i))) -
-                 distance(l, i);
+          eval = evalOld + transformationDiff * (norm(targetNeighbors(j, i)) +
+              norm(impostors(l, i)) + 2 * norm(i));
```

Comment:
Right, caching the norms for this bound can make a huge difference for speedup.

Reply:
Ya, these are cached norms (just have same name as arma::norm) :)